actionpack/lib/action_controller/metal/params_wrapper.rb

```diff
@@ -264,9 +264,11 @@ def _wrap_parameters(parameters)
       def _extract_parameters(parameters)
         if include_only = _wrapper_options.include
           parameters.slice(*include_only)
+        elsif _wrapper_options.exclude
+          exclude = _wrapper_options.exclude + EXCLUDE_PARAMETERS
+          parameters.except(*exclude)
         else
-          exclude = _wrapper_options.exclude || []
-          parameters.except(*(exclude + EXCLUDE_PARAMETERS))
+          parameters.except(*EXCLUDE_PARAMETERS)
         end
```

Comment:
I'm not sure if this is a hot spot worth optimizing, but we could tighten up the code a bit:

```ruby
else
  parameters.except(*_wrapper_options.exclude, *EXCLUDE_PARAMETERS)
end
```

(Note that `*nil` is equivalent to `*[]`.)

Reply:
Thanks for the suggestion! 
I went ahead and benchmarked it as well, but looks like that is even worse then the current implementation. Ruby allocates one more array than the current implementation, regardless of whether `_wrapper_options.exclude` is `nil` or not. I believe the reason is that in fact what ruby is doing for us is calling this:
```ruby
  parameters.except([*_wrapper_options.exclude, *EXCLUDE_PARAMETERS])
```
and `*nil` still becomes `[]`.

Reply:
Very interesting!  I had assumed each version would be roughly like `parameters.except([*EXCLUDE_PARAMETERS])`, because a callee can always safely mutate its splatted args.  But it appears that Ruby has a copy-on-write optimization when splatting a single array:

<details>
  <summary>Copy-on-write demonstration</summary>

  ```ruby
  require "benchmark/memory"

  def sort(*args)
    args.sort!
  end

  def pass(*args)
    args
  end

  EMPTY = []
  NONEMPTY = [3, 2, 1]

  Benchmark.memory do |x|
    x.report("sort nonempty") { sort(*NONEMPTY) }
    x.report("pass nonempty") { pass(*NONEMPTY) }
    x.report("sort empty") { sort(*EMPTY) }
    x.report("pass empty") { pass(*EMPTY) }
    x.report("sort empty * 2") { sort(*EMPTY, *EMPTY) }
    x.report("pass empty * 2") { pass(*EMPTY, *EMPTY) }
    x.compare!
  end
  ```
  ```
  Comparison:
         pass nonempty:         40 allocated
            sort empty:         40 allocated - same
            pass empty:         40 allocated - same
         sort nonempty:         80 allocated - 2.00x more
        sort empty * 2:        120 allocated - 3.00x more
        pass empty * 2:        120 allocated - 3.00x more
  ```
</details>

Additionally, it appears that Ruby does *not* have an optimization for splatting `nil`.  So while one version is like `parameters.except(EXCLUDE_PARAMETERS)`, the other is like `parameters.except([*nil.to_a, *EXCLUDE_PARAMETERS])`.  D'oh! :facepalm: